pkg/controller/machine_util.go

```diff
@@ -301,87 +347,161 @@ func (c *controller) syncMachineNodeTemplates(machine *v1alpha1.Machine) error {
-	mCopy, nCopy := machine.Spec.NodeTemplateSpec.Annotations, node.Annotations
-	updateNeeded := false
-
-	if mCopy == nil {
-		return false
+func SyncMachineLabels(
+	machine *v1alpha1.Machine,
+	node *v1.Node,
+	lastAppliedLabels map[string]string,
+) bool {
+	toBeUpdated := false
+	mLabels, nLabels := machine.Spec.NodeTemplateSpec.Labels, node.Labels
+
+	// Initialize node labels if nil
+	if nLabels == nil {
+		nLabels = make(map[string]string)
+		node.Labels = nLabels
 	}
-	if nCopy == nil {
-		nCopy = make(map[string]string)
+	// Intialize machine labels to empty map if nil
+	if mLabels == nil {
+		mLabels = emptyMap
 	}
 
-	for mkey, mvalue := range mCopy {
-		// if key doesnt exists or value does not match
-		if _, ok := nCopy[mkey]; !ok || mvalue != nCopy[mkey] {
-			updateNeeded = true
+	// Delete any labels that existed in the past but has been deleted now
+	for lastAppliedLabelKey := range lastAppliedLabels {
+		if _, exists := mLabels[lastAppliedLabelKey]; !exists {
+			delete(nLabels, lastAppliedLabelKey)
+			toBeUpdated = true
 		}
-		nCopy[mkey] = mvalue
 	}
-	if updateNeeded {
-		node.Annotations = nCopy
+
+	// Add/Update any key that doesn't exist or whose value as changed
+	for mKey, mValue := range mLabels {
+		if _, exists := nLabels[mKey]; !exists || mValue != nLabels[mKey] {
+			nLabels[mKey] = mValue
+			toBeUpdated = true
+		}
 	}
 
-	return updateNeeded
+	return toBeUpdated
 }
 
-// SyncMachineTaints syncs the annotations of the machine with node-objects.
+type taintKeyEffect struct {
+	// Required. The taint key to be applied to a node.
+	Key string
+	// Valid effects are NoSchedule, PreferNoSchedule and NoExecute.
+	Effect v1.TaintEffect
+}
+
+// SyncMachineTaints syncs the taints of the machine with node-objects.
 // It returns true if update is needed else false.
-func SyncMachineTaints(machine *v1alpha1.Machine, node *v1.Node) bool {
-	mTaintsCopy, nTaintsCopy := machine.Spec.NodeTemplateSpec.Spec.Taints, node.Spec.Taints
-	updateNeeded := false
-
-	for i := range mTaintsCopy {
-		elementFound := false
-		for j := range nTaintsCopy {
-			if mTaintsCopy[i] == nTaintsCopy[j] {
-				elementFound = true
-				break
-			} else if mTaintsCopy[i].Key == nTaintsCopy[j].Key {
-				elementFound, updateNeeded = true, true
-				nTaintsCopy[j] = mTaintsCopy[i]
-				break
-			}
-		}
-		if elementFound == false {
-			nTaintsCopy = append(nTaintsCopy, mTaintsCopy[i])
-			updateNeeded = true
+func SyncMachineTaints(
+	machine *v1alpha1.Machine,
+	node *v1.Node,
+	lastAppliedTaints []v1.Taint,
+) bool {
+	toBeUpdated := false
+	mTaints, nTaints := machine.Spec.NodeTemplateSpec.Spec.Taints, node.Spec.Taints
+	mTaintsMap := make(map[taintKeyEffect]*v1.Taint, 0)
+	nTaintsMap := make(map[taintKeyEffect]*v1.Taint, 0)
+
+	// Initialize node taints if nil
+	if nTaints == nil {
+		nTaints = []v1.Taint{}
+	}
+	// Intialize machine taints to empty map if nil
+	if mTaints == nil {
```

Comment:
I think range over nil slices is safe. Can we skip this?

Reply:
Done. Refer - https://github.com/gardener/machine-controller-manager/pull/268/commits/2d5030db1b9f88a560ddc8089446c9727145d135.